libp2p-floodsub/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+            .field("received", &self.received)
+            .finish()
+    }
+}
+
+impl<T, C> FloodSubController<T, C>
+where
+    T: MuxedTransport + 'static,                // TODO: 'static :-/
+    C: ConnectionUpgrade<T::RawConn> + 'static, // TODO: 'static :-/
+{
+    /// Builds a new controller for floodsub.
+    #[inline]
+    pub fn new(
+        upgrade: &FloodSubUpgrade,
+        swarm: SwarmController<T, C>,
+    ) -> FloodSubController<T, C> {
+        FloodSubController {
+            inner: upgrade.inner.clone(),
+            swarm: swarm,
+        }
+    }
+
+    /// Subscribe to a topic. When a node on the network sends a message for that topic, we will
+    /// likely receive it.
+    ///
+    /// It is not guaranteed that we receive every single message published on the network.
+    #[inline]
+    pub fn subscribe(&self, topic: &Topic) {
+        // This function exists for convenience.
+        self.subscribe_multi(iter::once(topic));
+    }
+
+    /// Same as `subscribe`, but subscribes to multiple topics at once.
+    #[inline]
+    pub fn subscribe_multi<'a, I>(&self, topics: I)
+    where
+        I: IntoIterator<Item = &'a Topic>,
+        I::IntoIter: Clone,
+    {
+        // This function exists for convenience.
+        self.sub_unsub_multi(topics.into_iter().map::<_, fn(_) -> _>(|t| (t, true)))
+    }
+
+    /// Unsubscribe from a topic. We will no longer receive any message for this topic.
+    ///
+    /// If a message was sent to us before we are able to notify that we don't want messages
+    /// anymore, then the message will be filtered out locally.
+    #[inline]
+    pub fn unsubscribe(&self, topic: &Topic) {
+        // This function exists for convenience.
+        self.unsubscribe_multi(iter::once(topic));
+    }
+
+    /// Same as `unsubscribe` but unsubscribes from multiple topics at once.
+    #[inline]
+    pub fn unsubscribe_multi<'a, I>(&self, topics: I)
+    where
+        I: IntoIterator<Item = &'a Topic>,
+        I::IntoIter: Clone,
+    {
+        // This function exists for convenience.
+        self.sub_unsub_multi(topics.into_iter().map::<_, fn(_) -> _>(|t| (t, false)));
+    }
+
+    // Inner implementation. The iterator should produce a boolean that is true if we subscribe and
+    // false if we unsubscribe.
+    fn sub_unsub_multi<'a, I>(&self, topics: I)
+    where
+        I: IntoIterator<Item = (&'a Topic, bool)>,
+        I::IntoIter: Clone,
+    {
+        let mut proto = rpc_proto::RPC::new();
+
+        let topics = topics.into_iter();
+
+        debug!(target: "libp2p-floodsub", "Queuing sub/unsub message ; sub = {:?} ; unsub = {:?}",
+               topics.clone().filter(|t| t.1)
+                     .map(|t| t.0.hash().clone().into_string())
+                     .collect::<Vec<_>>(),
+               topics.clone().filter(|t| !t.1)
+                     .map(|t| t.0.hash().clone().into_string())
+                     .collect::<Vec<_>>());
```

Comment:
If the `debug!` call does real work then it should be behind an `if in_debug_mode { ... }`